4 1 to 4 2 migration

apinnick opened the pull request:
[Formatted preview](http://file.tlv.redhat.com/~apinnick/4_1-to-4_2-migration/migration/migrating-4-4/migrating-openshift-4-to-4.html)

XinRedhat commented:
The source cluster should be `OpenShift Container Platform 4.1 `
>The Cluster Application Migration (CAM) tool enables you to migrate Kubernetes resources, persistent volume data, and internal container images from an OpenShift Container Platform 3.x source cluster to an OpenShift Container Platform 4.2 target cluster, using the CAM web console or the Kubernetes API.

XinRedhat commented:
In OCP 4.1, it can install the CAM Operator with OLM on the target cluster 
>Installing the CAM Operator manually on the source cluster

XinRedhat commented:
Can we clarify that `launching the CAM web console on the target cluster`? 
>4. Launching the CAM web console

apinnick commented:
> The source cluster should be `OpenShift Container Platform 4.1 `
> 
> > The Cluster Application Migration (CAM) tool enables you to migrate Kubernetes resources, persistent volume data, and internal container images from an OpenShift Container Platform 3.x source cluster to an OpenShift Container Platform 4.2 target cluster, using the CAM web console or the Kubernetes API.

Fixed.

apinnick commented:
> In OCP 4.1, it can install the CAM Operator with OLM on the target cluster
> 
> > 1. Installing the CAM Operator manually on the source cluster

Fixed

apinnick commented:
> Can we clarify that `launching the CAM web console on the target cluster`?
> 
> > 1. Launching the CAM web console

Fixed. I added a sentence to the first step (CAM Operator) to clarify that the UI is installed on the target cluster by default. However, I did not want to change this step because in the future we will support a migration with CAM installed on a non-target cluster.

apinnick commented:
Verified by QE

apinnick commented:
@bergerhoffer 

Can you review this? Thanks

Danil-Grigorev commented on migration/migrating-4-4/migrating-openshift-4-to-4.adoc:
```diff
@@ -0,0 +1,93 @@
+* Target cluster: Migration controller, CAM web console, Restic, and Velero
+
+. Configuring cross-origin resource sharing on the source cluster(s) to enable communication with the CAM tool
+:ocp-4_1-4_2!:
```
Seems like this tag is not properly resolved and is displayed as is.

djwhatle commented:
@apinnick reviewed the docs preview and identified a few more areas that will need changing. 

---

**Section: Adding a replication repository to the CAM web console**
We continue to support S3 API based replication repositories, but we now also support 2 more kinds of storage backends here: Azure Storage Containers and GCP Buckets.

When going through the Web UI flow of creating the replication repository, I believe there is now a drop-down menu allowing for selection of back-end type, and then the fields that you must fill out will change based on what is selected.

Each of these new backends is slightly different from S3 API compliant backend because they accept a _credentials blob_ instead of individual creds fields. This is due to the way that these items are handled by Velero.

 - PR with info on getting the credentials blob for Azure Storage Containers: https://github.com/fusor/mig-controller/pull/341
 - PR with info on getting the credential blob for GCP Storage Buckets

---

**Section: Creating a migration plan in the CAM web console**
There are now snapshot option selection steps in the "Migration Plan" creation wizard. We need a few more steps describing these new options in the docs.

djwhatle commented:
@apinnick round of screenshots coming up

---

### New `Azure cluster` checkbox during cluster creation requiring input of the "Resource Group" that all cluster resources belong to.

![image](https://user-images.githubusercontent.com/7576968/69350496-a5143e00-0c47-11ea-9b85-ff7cd8a9f8be.png)

---

![image](https://user-images.githubusercontent.com/7576968/69350476-9d549980-0c47-11ea-8e31-e51674d940b9.png)

---

### New first step when adding replication repository

![image](https://user-images.githubusercontent.com/7576968/69350749-19e77800-0c48-11ea-82e7-aa2d4c5074ae.png)

---

### AWS fields (stay the same)

![image](https://user-images.githubusercontent.com/7576968/69350811-31266580-0c48-11ea-93f0-98fb1c55acd6.png)

---

### GCP Fields (new form, needs to be documented)

![image](https://user-images.githubusercontent.com/7576968/69350905-65018b00-0c48-11ea-964a-238e325475d7.png)

---

### Azure fields (new form, needs to be documented)

![image](https://user-images.githubusercontent.com/7576968/69350941-78acf180-0c48-11ea-849c-cc367e6288ee.png)

djwhatle commented:
### Snapshot capability on Azure and GCP

![image](https://user-images.githubusercontent.com/7576968/69353172-17871d00-0c4c-11ea-9e7f-a98c0ec713a0.png)

Which results in...

```
apiVersion: migration.openshift.io/v1alpha1
kind: MigPlan
metadata:
  annotations:
    touch: db4df1dd-544d-4928-815f-1c6786627ce1
  creationTimestamp: 2019-11-21T15:43:04Z
  generation: 4
  name: my-plan
  namespace: openshift-migration
  resourceVersion: "22024709"
  selfLink: /apis/migration.openshift.io/v1alpha1/namespaces/openshift-migration/migplans/my-plan
  uid: 9b8d003e-0c75-11ea-8cdc-000d3a91f1c0
spec:
  destMigClusterRef:
    name: host
    namespace: openshift-migration
  migStorageRef:
    name: velero-s4
    namespace: openshift-migration
  namespaces:
  - mediawiki
  persistentVolumes:
  - capacity: 1Gi
    name: pvc-24ec0867-e6dc-11e9-ae9b-000d3a42e9fa
    pvc:
      accessModes:
      - ReadWriteOnce
      name: mediawiki
      namespace: mediawiki
    selection:
      action: copy
      copyMethod: snapshot
      storageClass: noobaa-storage-class
    storageClass: managed-premium
    supported:
      actions:
      - copy
      - move
      copyMethods:
      - filesystem
      - snapshot
  - capacity: 1Gi
    name: pvc-edc27824-e6db-11e9-ae9b-000d3a42e9fa
    pvc:
      accessModes:
      - ReadWriteOnce
      name: mariadb
      namespace: mediawiki
    selection:
      action: copy
      copyMethod: filesystem
      storageClass: managed-premium
    storageClass: managed-premium
    supported:
      actions:
      - copy
      - move
      copyMethods:
      - filesystem
      - snapshot
  srcMigClusterRef:
    name: azure2
    namespace: openshift-migration
```

apinnick commented:
@djwhatle has confirmed that the Azure and GCP comments do not belong to this PR.

djwhatle commented:
@apinnick hid the comments I made. Not sure where they belong yet but will copy them over at some point to an appropriate location.

bergerhoffer commented on migration/migrating-3-4/migrating-openshift-3-to-4.adoc:
```diff
@@ -1,7 +1,8 @@
 = Migrating {product-title} 3.x to 4.2
 include::modules/common-attributes.adoc[]
-:context: migrating-openshift-3-to-4
+:context: ocp-3-to-4
```
Is there a reason this context was changed? Typically we try not to change anchors once something has been published.

apinnick replied:
The context was too long and it seemed odd to see "migrating" so frequently in links. I think it's early enough in the publication process that changing the context won't be a problem. I'll keep it in mind for the future.

bergerhoffer commented on migration/migrating-3-4/migrating-openshift-3-to-4.adoc:
```diff
@@ -1,7 +1,8 @@
 include::modules/common-attributes.adoc[]
-:context: migrating-openshift-3-to-4
+:context: ocp-3-to-4
+:ocp-3-to-4:
```
What is this attribute for? I don't see it used in any ifdef/ifeval statements or anything.

bergerhoffer commented on migration/migrating-3-4/migrating-openshift-3-to-4.adoc:
```diff
@@ -49,19 +50,30 @@ The following `imagestreamtags` have been _removed_ from {product-title} 4.2:
 
-include::modules/migration-installing-migration-operator-manually.adoc[leveloffset=+2]
-include::modules/migration-installing-migration-operator-olm.adoc[leveloffset=+2]
+. Installing the CAM Operator on all clusters
```
Not critical, but just wondering if these should link to the appropriate below sections.

bergerhoffer commented on migration/migrating-3-4/migrating-openshift-3-to-4.adoc:
```diff
@@ -49,19 +50,30 @@ The following `imagestreamtags` have been _removed_ from {product-title} 4.2:
-// == Configuring cross-origin resource sharing
-//
-// You must configure cross-origin resource sharing on the source cluster to enable the CAM tool to communicate with the cluster's API server.
+:context: targetcluster3-4
```
I don't know that it's a good idea to change the value of the context attribute here in the middle of an assembly. Why is this being done?

If you need to use an attribute like the `targetcluster3-4` one you created below, you shouldn't need to touch the context here too.

bergerhoffer replied:
Update: I reached out to the rest of the team and they pointed me to the mod docs section that shows that this is what's needed to be done when a module is included multiple times in an assembly. You can ignore my questions on this. Though I do still have questions on a few  attributes that are defined but not ever used.

apinnick replied:
I wanted to use the same module for installing the CAM Operator on 4.1 source clusters and 4.2 target clusters.

bergerhoffer commented on migration/migrating-4-4/migrating-openshift-4-to-4.adoc:
```diff
@@ -0,0 +1,91 @@
+[id="migrating-openshift-4-to-4"]
+= Migrating {product-title} 4.1 to 4.2
+include::modules/common-attributes.adoc[]
+:context: ocp-4_1-4_2
```
Usually we do try to have the context match the anchor of the assembly (`migrating-openshift-4-to-4` in this case)

bergerhoffer commented on migration/migrating-4-4/migrating-openshift-4-to-4.adoc:
```diff
@@ -0,0 +1,91 @@
+= Migrating {product-title} 4.1 to 4.2
+include::modules/common-attributes.adoc[]
+:context: ocp-4_1-4_2
+:ocp-4_1-4_2:
```
Doesn't look like you're using this attribute anywhere

bergerhoffer commented on migration/migrating-4-4/migrating-openshift-4-to-4.adoc:
```diff
@@ -0,0 +1,91 @@
+
+:ocp-4_1-4_2!:
+
+:context: sourcecluster
```
Same thing here about changing the context not being something we should probably be doing.

bergerhoffer commented on modules/migration-installing-cam-operator-ocp-4.adoc:
```diff
@@ -32,6 +43,23 @@ The *Cluster Application Migration Operator* is listed in the *openshift-migrati
+ifdef::sourcecluster[]
+.. Update the `migration_controller` and `migration_ui` parameters and add the `deprecated_cors_configuration` parameter to the `spec` stanza:
++
+----
```
This should have source,yaml on the code block.

eriknelson commented on migration/migrating-3-4/migrating-openshift-3-to-4.adoc:
```diff
@@ -20,12 +21,12 @@ Before you begin your migration, be sure to review the information on xref:../mi
 * The target cluster must be {product-title} 4.2.
-* You must have xref:../cli_reference/openshift_cli/administrator-cli-commands.html#policy[`cluster-admin` privileges] on all clusters.
+* You must have xref:../../cli_reference/openshift_cli/administrator-cli-commands.html#policy[`cluster-admin` privileges] on all clusters.
 * You must have `podman` installed.
```
`podman` was really only required for 3.x clusters because you need to run and extract an `operator.yml` file off the container to install the operator. With 4.x clusters, OLM is installed with our metadata, so podman isn't required.

apinnick replied:
Good catch. That wasn't supposed to be there.

eriknelson commented on migration/migrating-3-4/migrating-openshift-3-to-4.adoc:
```diff
@@ -1,11 +1,12 @@
-You can migrate application workloads from {product-title} 3.7 (and later) to {product-title} 4.2 with the Cluster Application Migration (CAM) tool. The CAM tool enables you to control the migration and to minimize application downtime.
+You can migrate application workloads from {product-title} 3.7 (to 3.11) to {product-title} 4.2 with the Cluster Application Migration (CAM) tool. The CAM tool enables you to control the migration and to minimize application downtime.
 
 The CAM tool's web console and API, based on Kubernetes custom resources, enable you to migrate stateful application workloads at the granularity of a namespace.
```
The tool can perform stateful and stateless migrations.

kalexand-rh commented:
/cherrypick enterprise-4.2

kalexand-rh commented:
/cherrypick enterprise-4.3
